Migration from platform views 🚧

CoderNamedHendrick opened the pull request:
Story: #135 

## Summary

An attempt to migrate from embedding platform views and instead call the native navigation APIs. This update removes the need to use callbacks that return JSON strings in embedded platform views and simply sends type-safe responses to the Flutter application.

## Known Issues

There are no issues on Android.
On iOS, this change requires a new setup step, where the user needs to wrap the FlutterViewController with a UINavigationController in the app delegate. This is necessary to have page-to-page-like (the alternative is modal-like in the absence of the controller) navigation. Any ideas on how to work around this would be appreciated. This issue on iOS has been resolved see [here](https://github.com/smileidentity/flutter/pull/136/commits/81a157b67ee8e15b3a69c535ac6235be0e0650ad)

## Test Instructions

Run the sample app on both iOS and Android and test the selfie enrollment(the only flow currently migrated and awaiting feedback on implementation)

## Screenshot

If applicable (e.g. UI changes), add screenshots to help explain your work.
iOS and Android previews showing migrated smartSelfieEnrollment flows.

![ios_preview](https://github.com/user-attachments/assets/47ac6b44-6941-48f0-9335-c09b465161ab)
![android_preview](https://github.com/user-attachments/assets/bd16e1a8-c4fa-433c-baa4-f77aac00f406)

CoderNamedHendrick commented:
@jumaallan I'll be awaiting your feedback

jumaallan commented on android/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1 @@
+<resources></resources>
```
Do we need this?

CoderNamedHendrick replied:
No we don't, android studio generated it. I'll remove it

jumaallan commented on example/ios/Podfile.lock:
```diff
@@ -48,4 +48,4 @@ SPEC CHECKSUMS:
 PODFILE CHECKSUM: 929954fb8941cef06249e96bd1516fd2a22ed7a5
 
-COCOAPODS: 1.16.2
+COCOAPODS: 1.15.2
```
this is a lower version

jumaallan commented on ios/Classes/SmileIDSmartSelfieEnrollmentView.swift:
```diff
@@ -0,0 +1,67 @@
+//
+//  SmileIDSmartSelfieEnrollmentView.swift
```
we usually remove these file comments

jumaallan commented on pigeon/messages.dart:
```diff
@@ -804,21 +837,11 @@ class FlutterConfig {
-      bool useSandbox,
-      bool enableCrashReporting
-  );
+      String apiKey, FlutterConfig config, bool useSandbox, bool enableCrashReporting);
```
we have no linter, but since the arguments are long, we can use separate them in different lines

jumaallan commented on pigeon/messages.dart:
```diff
@@ -804,21 +837,11 @@ class FlutterConfig {
-      bool useSandbox,
-      bool enableCrashReporting
-  );
+  void initializeWithConfig(FlutterConfig config, bool useSandbox, bool enableCrashReporting);
```
same here

jumaallan commented on pigeon/messages.dart:
```diff
@@ -835,8 +858,11 @@ abstract class SmileIDApi {
-  FlutterAuthenticationResponse authenticate(
-      FlutterAuthenticationRequest request);
+  SmartSelfieCaptureResult smartSelfieEnrollment(
+      SmartSelfieEnrollmentCreationParams creationParams);
```
I prefer the closing `);` on a new line

jumaallan commented on android/build.gradle:
```diff
@@ -106,3 +106,10 @@ ktlint {
         exclude { it.file.path.contains(".g.kt") }
     }
 }
+
```
remove these dependencies too

jumaallan replied:
I fixed this for you

jumaallan commented on android/src/main/AndroidManifest.xml:
```diff
@@ -1 +1,10 @@
+    <application>
+        <activity
+                android:name=".SmileIDSmartSelfieEnrollmentActivity"
+                android:exported="false" />
```
we might want to export this - set it to `true`

jumaallan replied:
I fixed this for you

jumaallan commented on example/lib/main.dart:
```diff
@@ -195,11 +195,55 @@ class MainContent extends StatelessWidget {
+
+        print('enrollment selfie: ${result.selfieFile}');
+        print('enrollment liveness: ${result.livenessFiles}');
+        print('enrollment apiResponse: ${result.apiResponse}');
```
I can't see the apiResponse and didSubmitBiometricKycJob responses

not sure why we log or return the didSubmitBiometricKycJob on a non biometric job too

CoderNamedHendrick replied:
I don't think didSubmitBiometricKycJob is returned from the call, I only added it since it is part of the response object, I did a one-to-one mapping there, I'll remove it.

I'll have to confirm the API response, I couldn't do so earlier today due to some issues with my smile id config

jumaallan replied:
alright, if you need any help I can take a look at fixing it

jumaallan commented on example/lib/main.dart:
```diff
@@ -195,11 +195,55 @@ class MainContent extends StatelessWidget {
+      //           ScaffoldMessenger.of(context).showSnackBar(snackBar);
+      //           Navigator.of(context).pop();
+      //         },
+      //         onError: (String errorMessage) {
```
we had the `onSuccess` and `onError` callbacks - so how do we handle errors now?

CoderNamedHendrick replied:
We could throw SmileIDException which would extend platform exception. Going with this approach we can have a function that takes the standard platform exception and throws a smileIDExcpetion to communicate these exceptions to the client.
```dart
try {
     final result = await SmileID.api.smartSelfieEnrollment(
          SmartSelfieEnrollmentCreationParams(
              allowNewEnroll: false,
              allowAgentMode: false,
              showAttribution: true,
              showInstructions: true,
              skipApiSubmission: false),
        );
} on SmileIdException catch (e) {
     // do something with exception
}
```

Another way would be creating a union type with sealed classes, where we can either return a success with an object(map this to a generic) or an error with the error message. Going with this approach we can ensure the front-facing APIs don't require try-catch blocks and they can be handled with dart pattern-matching which reads better to be fair. An example would look like this
```dart
 final result = await SmileID.api.smartSelfieEnrollment(
          SmartSelfieEnrollmentCreationParams(
              allowNewEnroll: false,
              allowAgentMode: false,
              showAttribution: true,
              showInstructions: true,
              skipApiSubmission: false),
        );

switch(result){
    case Success(: final data):
        // do something
    case Error(: final errorMessage):
        // do something on error
}
```

Which would you prefer? Any alternative implementation?

jumaallan replied:
I like the union type since the api already works like this - this also avoids the try catch like you said, so we can go with this approach

jumaallan commented:
I also noticed since we now use an activity, we now have a topbar added - not sure if we want to keep that or add a theme to hide it 

see these two screenshots

![Screenshot_20250108_220914](https://github.com/user-attachments/assets/d1e937f5-6ca3-48e1-aac5-7f472702f19c)
![Screenshot_20250108_220920](https://github.com/user-attachments/assets/01ea2937-5a41-4ee6-9350-acea5e489c25)


I think attribution does not work too, we might need to see why that is broken (maybe the screen content gets pushed way down)

CoderNamedHendrick commented:
> I also noticed since we now use an activity, we now have a topbar added - not sure if we want to keep that or add a theme to hide it
> 
> see these two screenshots
> 
> ![Screenshot_20250108_220914](https://private-user-images.githubusercontent.com/25085146/401258899-d1e937f5-6ca3-48e1-aac5-7f472702f19c.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYzNjE0NjEsIm5iZiI6MTczNjM2MTE2MSwicGF0aCI6Ii8yNTA4NTE0Ni80MDEyNTg4OTktZDFlOTM3ZjUtNmNhMy00OGUxLWFhYzUtN2Y0NzI3MDJmMTljLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA4VDE4MzI0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJkYmQzN2JjYTUxMjlmZWU0MGRmYThlMzEzMjBkNTBiODNjNTZhZTg0ODRjNDJiY2MyNTM2NjBkYzM2ODM1MTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.bjAPVNrxPe4al0w36zq5yO-CSD5S9TwysmAhu8cYHmU) ![Screenshot_20250108_220920](https://private-user-images.githubusercontent.com/25085146/401259059-01ea2937-5a41-4ee6-9350-acea5e489c25.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYzNjE0NjEsIm5iZiI6MTczNjM2MTE2MSwicGF0aCI6Ii8yNTA4NTE0Ni80MDEyNTkwNTktMDFlYTI5MzctNWE0MS00ZWU2LTkzNTAtYWNlYTVlNDg5YzI1LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA4VDE4MzI0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkZWFlZDc5ZDM5Yjk1NDI2ZGZiODc5OTcwMzg3MzM3ODE5MWI2M2M3ZTcyOTBlYjhlNGE5N2UxNTEwOWZkZjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.wFDjkOQY0zxqefdW-9E2SP00j4Zf0Oz5E_ZNuJMneMg)
> 
> I think attribution does not work too, we might need to see why that is broken (maybe the screen content gets pushed way down)

I just tested this and the show attribution is working on my device, you are right about the screen content being pushed down. Hiding the top bar should resolve this right? The top bar being there or not doesn't make much of a difference, most users of the current platform views simply follow the smileID docs and place a topbar with smile ID as the title :)

jumaallan commented:
> d resolve this right? The top bar bei

Yeah, go ahead and remove it then - I suspected the attribution was being pushed down the screen

CoderNamedHendrick commented:
I'll update the PR before the end of today to get your feedback, @jumaallan. Thanks for the quick responses! 😊

CoderNamedHendrick commented on example/ios/Runner.xcodeproj/project.pbxproj:
```diff
@@ -474,7 +474,7 @@
 				CODE_SIGN_STYLE = Automatic;
 				CURRENT_PROJECT_VERSION = "$(FLUTTER_BUILD_NUMBER)";
-				DEVELOPMENT_TEAM = 99P7YGX9Q6;
+				DEVELOPMENT_TEAM = 5SB38K2R24;
```
I had to use my company's development account. This will be reverted before merging. I'll keep this comment open so I don't forget to revert it once it's complete.

CoderNamedHendrick commented on example/ios/Flutter/AppFrameworkInfo.plist:
```diff
@@ -21,6 +21,6 @@
   <string>1.0</string>
   <key>MinimumOSVersion</key>
-  <string>11.0</string>
+  <string>12.0</string>
```
I believe `flutter run` made this change. Also this was done on the example app which doesn't affect applications where the plugin is integrated

CoderNamedHendrick commented on example/lib/main.dart:
```diff
@@ -195,11 +196,59 @@ class MainContent extends StatelessWidget {
+            print('error occurred with smart selfie enrollment: $error');
+        }
+      },
+      // onPressed: () {
```
Leaving this open for when the PR is complete

CoderNamedHendrick commented on android/src/main/kotlin/com/smileidentity/flutter/SmileIDPlugin.kt:
```diff
@@ -443,13 +475,57 @@ class SmileIDPlugin :
+    override fun onDetachedFromActivity() {
+    }
+
+    override fun onActivityResult(
```
Currently looking at the feasibility of implementing this

CoderNamedHendrick replied:
@JNdhlovu I am not sure how feasible this is. The activity that the ActivityAware interface provides doesn't have methods that work to register an activity result. Do you have a sample using it in a flutter plugin?

CoderNamedHendrick replied:
I can confirm that flutter isn't using androidx so is unable to use this API. @JNdhlovu

CoderNamedHendrick commented on android/src/main/kotlin/com/smileidentity/flutter/SmileIDPlugin.kt:
```diff
@@ -173,6 +181,30 @@ class SmileIDPlugin :
         }
     }
 
+    override fun smartSelfieEnrollment(
```
Hmmmmm, MethodChannel.MethodCallHandler isn't needed since we're implementing an interface Pigeon created. I do agree with this concern and have been investigating making a separate pigeon message file for the SDK products specifically. I'll be trying that out and updating the PR as soon as I look into the feasibility of implementing your earlier comment.

CoderNamedHendrick commented:
Added 3 more smile id products; smart selfie authentication, enhanced enrollment and enhanced authentication.

@JNdhlovu I moved calls to the products' APIs to its implementation class which should resolve your earlier concern, do take a look. I'll be expecting your valuable feedback.

CoderNamedHendrick commented:
@JNdhlovu I've refactored some of the setup, now client applications do not need to wrap their apps in a UINavigationController🚀

tobitech commented on ios/Classes/SmileIDProductsPluginApi.swift:
```diff
@@ -0,0 +1,81 @@
+        
+        if let controller = window??.rootViewController as? UINavigationController {
+            api.navigationController = controller
+            return;
```
semicolon not really needed here.

tobitech commented:
Tested and Jobs run fine on iOS.

CoderNamedHendrick commented on android/src/main/AndroidManifest.xml:
```diff
@@ -1 +1,22 @@
+
+    <application android:theme="@android:style/Theme.Light.NoTitleBar">
+        <activity
+                android:name=".SmileIDSmartSelfieEnrollmentActivity"
```
Yeah, I'll lock the implementation and update the PR later in the week while migrating the other flows

jumaallan commented:
@CoderNamedHendrick some iOS notes
* We need to lock the iOS orientation to portrait (on landscape the preview is half dark)
* We need to force the theme to light mode - theming is not 100% working on iOS so we normally lock it to light mode (refer to this PR - https://github.com/smileidentity/ios/pull/219)

CoderNamedHendrick commented on android/src/main/kotlin/com/smileidentity/flutter/SmileIDProductsPluginApi.kt:
```diff
@@ -0,0 +1,235 @@
+        intent.putSmartSelfieCreationParams(creationParams)
+
+        smartSelfieResult = callback
+        activity?.startActivityForResult(
```
I'll handle that. Funny I managed it on iOS but didn't think much about it on android. Thanks for point it out

CoderNamedHendrick commented:
> @CoderNamedHendrick some iOS notes
> 
> * We need to lock the iOS orientation to portrait (on landscape the preview is half dark)
> * We need to force the theme to light mode - theming is not 100% working on iOS so we normally lock it to light mode (refer to this PR - [Temporarily force the light mode even if the user is in dark mode ios#219](https://github.com/smileidentity/ios/pull/219))

Yeah @jumaallan I noticed this dark mode behaviour and forcing it to light mode is the way to go, thanks for the tip. I'll also lock the orientation in my next update.

CoderNamedHendrick commented on android/src/main/kotlin/com/smileidentity/flutter/SmileIDProductsPluginApi.kt:
```diff
@@ -0,0 +1,235 @@
+    errorCode: String,
+    resultCallback: (Result<SmartSelfieCaptureResult>) -> Unit,
+) {
+    if (resultCode == Activity.RESULT_OK) {
```
Oddly, I'm not able to replicate this. Are you looking at the logs from a Flutter run or did you run the app from the plugin on Android Studio? Doing so hides the logs printed by flutter

![image](https://github.com/user-attachments/assets/17240806-0aef-4697-8ebd-516e38e4decf)

CoderNamedHendrick commented on ios/Classes/SmileIDProductsPluginApi.swift:
```diff
@@ -0,0 +1,81 @@
+
+public class SmileIDProductsPluginApi: SmileIDProductsApi {
+    
+    var navigationController: UINavigationController? = nil
```
Concerning iOS, locking orientation is enforced on an app level and my attempts at locking the orientation of a single view weren't successful. The other attempts which influence the app-wide orientation state are stated [here](https://stackoverflow.com/questions/66037782/swiftui-how-do-i-lock-a-particular-view-in-portrait-mode-whilst-allowing-others) but need to evaluate if we need to do so.

I also observed in the earlier implementation the orientation wasn't fixed so do we leave setting the orientation to portrait on the developer as is currently done? @jumaallan @JNdhlovu

jumaallan commented:
@JNdhlovu @tobitech @robin-ankele 

There are more changes on this PR, so please take a look at this again

CoderNamedHendrick commented:
@jumaallan I have some changes I'm currently working on that I'm yet to push. As soon as I finish within the week, it can be reviewed.

CoderNamedHendrick commented:
Hey Guys, @jumaallan @JNdhlovu. Sorry I couldn't push the remaining update earlier, it got quite busy at work.

I've attended to the remaining products now and will be awaiting feedback from the team. Thanks

jumaallan commented:
Just 2 issues on iOS, otherwise this looks good to me @CoderNamedHendrick 
1. We still need to lock it on portrait mode
2. The modal has the back button placed too close to the modal (see this screenshots)


Since this is going to merge to an epic branch, I will approve it to unblock the next refactors for you as we fix these two minor issues

Thank you so much for helping us make this Flutter SDK better ❤️ 


![IMG_0080](https://github.com/user-attachments/assets/c746deba-4626-49ea-a1d0-65c42d06bbf1)
 
![IMG_0081](https://github.com/user-attachments/assets/ebba4aa6-3da9-4e4e-a34c-f52166583300)

jumaallan reviewed (approved):
Thank you for helping make this Flutter SDK better ❤️
